### DISCUSSION TITLE
Bigquad + servo

### DISCUSSION THREAD

**Discussion by Eric-Ewing** [bitcraze/discussions#1005] [Q&A] [answered]
Hello,

We've been working on adding a servo motor to a bigquad and are running into an issue that is proving hard for us to debug. The servo works when not in flight and it also works the first time you call it in flight, but the second time its called the motors shut off. 

We wrote a separate [servo.c driver](https://github.com/ewinge1/crazyflie-firmware/blob/4f7051b638662305a3a8264bb7e75cddc4fbbb4f/src/drivers/src/servo.c), that can move the servo to a desired angle. We've wired the servo to get signal from the cppm port on the bigquad deck. It uses TIM14, which we thought is currently unused. In our code we use the const servoMapMOSI, which is just set to be `const MotorPerifDef* servoMapMOSI = &MOTORS_PA7_TIM14_CH1_BRUSHLESS_OD;` in motors_def.c. We then use servoMapMosi in servo.c to setup our timer and gpio settings.

We can adjust the angle both through cfclient and crazyswarm by updating the parameters and it works fine on the ground (you can update it as many times as you want). We've built it to drop paint onto a canvas, so the normal pipeline is to takeoff, fly to the canvas, open the servo, do some stuff over the canvas, close the servo and return to a landing position. We can reliably do everything up to closing the servo. When the second call to servoSetAngle takes place, the motors all cut power and the bigquad free falls. You can see that happen here: https://youtu.be/szdlBCcJfyc?feature=shared, where it works fine up until the point we try to close the servo (unfortunately, directly over the canvas). It's hard to see the paint falling, but you can see orange start to show up on the canvas when we open the servo, and it doesn't show up before then.

What we've done to debug/test:

1. unplugging the servo, but running the same code produces the same behavior, so it's not an issue with too much current being drawn by the servo.
2. Taking the propellers off and attempting to run on the ground produces the same behavior (motors cut after second call to servo).
3. Removing the setCompare from the servoSetAngle method makes the motors not cut (although the servo obviously doesn't do anything).

So the questions we have are:
1. Is there something inherently wrong with using TIM14 for this task?
2. Is there something obviously wrong with our setCompare or timer use? (we're by no means embedded programmers)
3. Do you have any other suggestions for things we should test?

I probably left out some important info, so just let me know if you have any questions.

**Comment by tobbeanton** [comment]
Reviewing the code I can not see any direct problem, it should work... The [debug print](https://github.com/ewinge1/crazyflie-firmware/blob/4f7051b638662305a3a8264bb7e75cddc4fbbb4f/src/drivers/src/servo.c#L128) might be an issue though and could be a bad idea to do in the paramter update callback.

**Reply by tobbeanton** [reply]
And the debug print in saturateAngle as well.

**Reply by Eric-Ewing** [reply]
Yeah, it definitely has us stumped as well. Upon further experimentation, it seems to occur primarily while decreasing the angle (and thus the compare register). We can usually increase the angle (and thus the compare value) with something like:

```    
# Start "flight" somehow, either with motorParamSet or cf.takeoff()

for i in range(6):
        cf.setParam('servo.servoAngle', i*5)
        timeHelper.sleep(2.)
```
And it will behave as expected.

But as soon as we try to decrease the angle by any amount, the main motors cut out.
```    
for i in range(5):
        cf.setParam('servo.servoAngle', (5-i)*5)
        timeHelper.sleep(2.)
```
This code makes motors cut when the angle is decreasing. The servo however, does continue to work and will step back to 0, just with the main motors off.

But also sometimes increasing the angle doesn't work either and it cuts the motors.

We've tried it with a couple different crazyflies (one was 2.1 and one 2.0) and had the same results on both.

Would looking at it with a JTAG debugger be worthwhile? At this point I'm debating between investing more time looking at the current code vs. just adding an arduino or other dedicated servo controller to the drone. We have 4 total that are in various phases of being built and would like to offer support for servos in general in a art+robots class we're offering this spring, so it would be nice to avoid requiring an arduino for each bigquad.

**Reply by tobbeanton** [reply]
I'll have a look at it today and see if I can replicate it and hopefully find the issue.

**Reply by tobbeanton** [reply]
I did not have an oscilloscope with me today so have to finish tomorrow but I'm pretty hopeful I can fix it.

**Reply by tobbeanton** [reply, marked as answer]
I created a [pull-request](https://github.com/ewinge1/crazyflie-firmware/pull/3) to your branch which hopefully will fix your issue.

**Reply by Eric-Ewing** [reply]
This patch fixed our bug as far as I can tell, thanks so much! This is a pretty quick way to attach a servo in addition to the 4 main motors as long as you don't need cppm, as you can solder connectors to the cppm ports and plug the standard servo connectors directly to those pins. I've seen a few other discussions where this might be helpful, e.g. #988

**Reply by matejkarasek** [reply]
Awesome!
Would be great to have such "servo deck" pulled into bitcraze master, it could support up to 4 servos e.g. via the motor connectors of the bigquad deck... @tobbeanton what would still need to be changed?

**Reply by tobbeanton** [reply]
I agree. If it is OK with you @Eric-Ewing I will modify the code to make it a bit more generic and push it to master. (if i get time...)

**Reply by Eric-Ewing** [reply]
Sounds great. I think it'll lead to some people put together some cool projects with additional actuators on their bigquads or bolts.

**Reply by matejkarasek** [reply]
Here the PR for the servo deck: https://github.com/bitcraze/crazyflie-firmware/pull/1340

**Comment by matejkarasek** [comment]
I did not read your question and the answer of Tobias in detail, but in case the proposed solution would not work you can also check the "flapper" platform that we help maintain.
It uses 2 brushless motors and 2 servos, maybe you can find some inspiration there? It is Bolt based, but we do have a branch that allows extra servomotors (3+), those are connected Big-quad style... A bit hacky, but I can share if needed.

**Reply by Eric-Ewing** [reply]
Thanks for the pointers, we did look a bit at the flapper code when trying to write our servo firmware. It ultimately ended up being a bug with some of our pin settings that was leading to buggy behavior.

**Comment by matejkarasek** [comment]
For anyone still looking for this functionality, it is now in master: https://github.com/bitcraze/crazyflie-firmware/pull/1340
